xbuild/src/command/build.rs

```diff
@@ -261,9 +261,67 @@ pub fn build(env: &BuildEnv) -> Result<()> {
+                    let cargo_dir = arch_dir.join("cargo");
+                    let lib = env.cargo_artefact(&cargo_dir, target, CrateType::Cdylib)?;
                     apk.add_lib(target.android_abi(), &lib)?;
+
+                    let ndk = env.android_ndk();
+
+                    let deps_dir = {
+                        let arch_dir = if target.is_host()? {
+                            cargo_dir.to_path_buf()
+                        } else {
+                            cargo_dir.join(target.rust_triple()?)
+                        };
+                        let opt_dir = arch_dir.join(target.opt().to_string());
+                        opt_dir.join("deps")
+                    };
+
+                    let mut search_paths = env
+                        .cargo()
+                        .lib_search_paths(&cargo_dir, target)
+                        .with_context(|| {
+                            format!(
+                                "Finding libraries in `{}` for {:?}",
+                                cargo_dir.display(),
+                                target
+                            )
+                        })?;
+                    search_paths.push(deps_dir);
+                    let search_paths = search_paths.iter().map(AsRef::as_ref).collect::<Vec<_>>();
+
+                    let ndk_sysroot_libs = ndk.join("usr/lib").join(target.ndk_triple());
+                    let provided_libs_paths = [
+                        ndk_sysroot_libs.as_path(),
+                        &*ndk_sysroot_libs.join(
+                            // Use libraries (symbols) from the lowest NDK that is supported by the application,
+                            // to prevent inadvertently making newer APIs available:
+                            // https://developer.android.com/ndk/guides/sdk-versions
+                            env.manifest()
+                                .android()
+                                .sdk
+                                .min_sdk_version
+                                .unwrap()
+                                .to_string(),
+                        ),
+                    ];
+
+                    let extra_libs = xcommon::llvm::list_needed_libs_recursively(
+                        &lib,
+                        &search_paths,
+                        &provided_libs_paths,
+                    )
+                    .with_context(|| {
+                        format!(
+                            "Failed to collect all required libraries for `{}` with `{:?}` available libraries and `{:?}` shippable libraries",
+                            lib.display(),
+                            provided_libs_paths,
+                            search_paths
+                        )
+                    })?;
+                    for lib in &extra_libs {
+                        apk.add_lib(target.android_abi(), lib)?;
+                    }
```

Comment:
the build.rs file is getting quite large. maybe we can have a file for each platform handling the final packaging.

Reply:
Yes, that's what I was hinting at when mentioning it being hard to find the right pieces of code to edit.

I don't want to do that as part of this PR though to make it unnecessarily larger. Feel free to pick it up after merging, or before and I'll rebase this on top.